src/cargo/util/toml/mod.rs

```diff
@@ -2561,6 +2592,17 @@ impl<P: ResolveToPath + Clone> DetailedTomlDependency<P> {
     fn update_optional(&mut self, optional: Option<bool>) {
         self.optional = optional;
     }
+
+    fn resolve_path(&mut self, root_path: &Path, package_root: &Path, config: &Config) {
+        if let Some(from_root) = &self.path {
+            let resolved = from_root.resolve(config);
+            if let Some(diff_path) = diff_paths(root_path.join(resolved), package_root) {
+                if let Some(diff_str) = diff_path.to_str() {
+                    self.path = Some(diff_str.to_owned());
+                }
+            }
```

Comment:
If I'm reading this correctly, we are leaving the incorrect relative path in if it doesn't work?

Wonder if we should have a function shared between the the path rewriting so we can be consistent on error reporting.

Reply:
You mean something like
```rust
pub fn resolve_relative_path(label: &str, root_path: Pathbuf, package_root: PathBuf) -> CargoResult<String> {
    match diff_paths(root_path, package_root) {
        None => Err(anyhow!("`workspace.{}` was defined but could not be resolved with {}", label, package_root.display())),
        Some(path) => Ok(
            path
            .to_str()
            .ok_or_else(|| anyhow!("`{}` resolved to non-UTF value (`{}`)", label, path.display()))?
            .to_owned()
        ),
    }
}
```

Reply:
- I'd recommend having `old_root`, `new_root`, and `rel_path` as parameters so it can also handle `join` + `normalize_path` for you
- `workspace.` doesn't quite work for path dependencies.  Maybe just leave it as `{label}` and re-word to the first to include both paths

Reply:
```rust
pub fn resolve_relative_path(
    label: &str,
    old_root: &PathBuf,
    new_root: &PathBuf,
    rel_path: &str,
) -> CargoResult<String> {
    let joined_path = normalize_path(&old_root.join(rel_path));
    match diff_paths(joined_path, new_root) {
        None => Err(anyhow!(
            "`{}` was defined in {} but could not be resolved with {}",
            label,
            old_root.display(),
            package_root.display()
        )),
        Some(path) => Ok(path
            .to_str()
            .ok_or_else(|| {
                anyhow!(
                    "`{}` resolved to non-UTF value (`{}`)",
                    label,
                    path.display()
                )
            })?
            .to_owned()),
    }
}
```

Reply:
Seems like that'd work